folding-schemes/src/folding/hypernova/decider_eth_circuit.rs

```diff
@@ -0,0 +1,627 @@
+        let incircuit_challenge =
+            KZGChallengeGadget::<C1>::get_challenge_gadget(&mut transcript, U_i1.clone())?;
+        incircuit_challenge.enforce_equal(&kzg_challenge)?;
+
+        // 6.b check that the obtained U_{i+1} from the NIMFS.V matches the U_{i+1} from the input,
+        // except for the C parameter, which to compute its folding would require non-native logic
+        // in-circuit, and we defer it to outside the circuit.
+        computed_U_i1.u.enforce_equal(&U_i1.u)?;
+        computed_U_i1.r_x.enforce_equal(&U_i1.r_x)?;
+        computed_U_i1.v.enforce_equal(&U_i1.v)?;
+
+        // 8.b check that the in-circuit computed r is equal to the inputted r.
+        // Notice that rho_powers only contains one element, since at the final fold we are only
+        // folding 2-to-1 instances.
+        let rho_powers: Vec<FpVar<CF1<C1>>> = rho_vec
+            .iter()
+            .map(|rho_i| Boolean::le_bits_to_fp_var(rho_i))
+            .collect::<Result<_, SynthesisError>>()?;
+
+        let external_rho_powers = Vec::<FpVar<CF1<C1>>>::new_input(cs.clone(), || {
+            Ok(self.rho_powers.unwrap_or(vec![CF1::<C1>::zero(); 1]))
+        })?;
+        rho_powers.enforce_equal(&external_rho_powers)?;
+
+        Ok(())
+    }
+}
+
+/// Gadget that computes the KZG challenges, also offers the rust native implementation compatible
+/// with the gadget.
+pub struct KZGChallengeGadget<C: CurveGroup> {
+    _c: PhantomData<C>,
+}
+#[allow(clippy::type_complexity)]
+impl<C> KZGChallengeGadget<C>
+where
+    C: CurveGroup,
+    C::ScalarField: PrimeField,
+    <C as CurveGroup>::BaseField: PrimeField,
+    C::ScalarField: Absorb,
+{
+    pub fn get_challenge_native<T: Transcript<C::ScalarField>>(
+        transcript: &mut T,
+        U_i: LCCCS<C>,
+    ) -> Result<C::ScalarField, Error> {
+        // compute the KZG challenges, which are computed in-circuit and checked that it matches
+        // the inputted one
+        transcript.absorb_nonnative(&U_i.C);
+        Ok(transcript.get_challenge())
+    }
+    // compatible with the native get_challenge_native
+    pub fn get_challenge_gadget<S: CryptographicSponge, T: TranscriptVar<CF1<C>, S>>(
+        transcript: &mut T,
+        U_i: LCCCSVar<C>,
+    ) -> Result<FpVar<C::ScalarField>, SynthesisError> {
+        transcript.absorb(&U_i.C.to_constraint_field()?)?;
+        transcript.get_challenge()
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use ark_bn254::{constraints::GVar, Fr, G1Projective as Projective};
+    use ark_grumpkin::{constraints::GVar as GVar2, Projective as Projective2};
+    use ark_relations::r1cs::ConstraintSystem;
+    use ark_std::One;
+    use ark_std::{test_rng, UniformRand};
+
+    use super::*;
+    use crate::commitment::pedersen::Pedersen;
+    use crate::folding::nova::PreprocessorParam;
+    use crate::frontend::tests::CubicFCircuit;
+    use crate::transcript::poseidon::poseidon_canonical_config;
+    use crate::FoldingScheme;
+
```

Comment:
Unsure I follow this comment. What's `i` here?

Reply:
That was not clear indeed! I meant to test it when the number of IVC iteration is at 0.

Reply:
I've done it locally, and it adds ~15 extra seconds to the test, we would do it also for the Nova's decider_eth_circuit test, so it would be ~30 extra seconds of tests.
On the same time, I'm not sure how realistic is generating a decider proof for a folding scheme prior to do any fold. If somebody wants to do that probably they would be better just doing a proof without folding directly.